### PR TITLE
windows-exe: optional sections in silent (/S) mode, bug fixes

### PR DESCRIPTION
**Pull request by inikulshin**
1. Command line controlled optional sections (Start Menu, Service Install, Service Run) - `/NOSTARTMENU` `/NOSERVICEINSTALL` `/NOSERVICERUN` command line flags
2. Move "Start RabbitMQ service" step to final window.
3. Check rabbitmq-service.bat exit codes (and fail on error).
4. Use `RABBITMQ_BASE`, `RABBITMQ_LOG_BASE`, `RABBITMQ_MNESIA_BASE` and `RABBITMQ_PLUGINS_DIR` from environment - to put files (plugins, config) in right place and create right shortcuts)
5. Fix Erlang version check (compare right versions interval)
6. `Unicode false` - for makensis.exe >=3.07

### CODE REVIEW

**Comment by inikulshin** [comment]
The changed file is **identical** in all other branches (3.8.x, 3.9.x, 3.10.x) except 2 Pivotal/VMWare diffs in 3.8.x (which also should be fixed, imho).

So, after this one will be approved, it can (and should) be easily picked to all other branches. The only difference is

master, 3.10.x:
`!define ERLANG_TO_VERSION   "13.1"      ; https://www.erlang.org/patches/otp-25.0#erts-13.0`

3.8.x, 3.9.x:
`!define ERLANG_TO_VERSION   "12.4"  ; https://www.erlang.org/patches/otp-24.3.4#erts-12.3.2`

**Comment by lukebakken** [comment]
Thank you! I'll find time within the next week or two to build and test an installer with these changes.

**Comment by inikulshin** [comment]
@lukebakken and what about other branches?
Will you cherry pick once this PR is approved?

Or you prefer me to prepare 3 more almost identical PRs **already**?
Our primary goal is to get those changes in 3.8.x and use them in the new release (**3.8.35**?).

**Comment by michaelklishin** [comment]
We will take care of back porting, no need to submit multiple identical PRs.

**Comment by inikulshin** [comment]
@lukebakken @michaelklishin can you schedule PR review and testing, please?
I understand that it has very low priority, but it's almost done.

**Review by lukebakken — CHANGES_REQUESTED** [review]
Here is how I built the installer:

```
cd development/rabbitmq/rabbitmq-server

lbakken@shostakovich ~/development/rabbitmq/rabbitmq-server (master =)
$ ls ../rabbitmq-packaging/
common  debs  Makefile  RPMS  windows  windows-exe
```

Note that the above clone is a branch with your `master` branch merged in.

Here is the command I ran and the end of the output

```
make FULL=1 RABBITMQ_PACKAGING_REPO="/home/lbakken/development/rabbitmq/rabbitmq-packaging" UNIX_TO_DOS=unix2dos package-windows
```

End of output:

```
make -C windows-exe SOURCE_DIST_FILE="/home/lbakken/development/rabbitmq/rabbitmq-server/PACKAGES/rabbitmq-server-3.8.10+1443.gf95d2ba.tar.xz" PACKAGES_DIR="/home/lbakken/development/rabbitmq/PACKAGES" SIGNING_KEY="" all clean
make[2]: Entering directory '/home/lbakken/development/rabbitmq/rabbitmq-packaging/windows-exe'
sed \
    -e 's|%%VERSION%%|3.8.10+1443.gf95d2ba|' \
    -e 's|%%PRODUCT_VERSION%%|3.8.10.0|' \
    rabbitmq_nsi.in > rabbitmq-3.8.10+1443.gf95d2ba.nsi
unzip -q /home/lbakken/development/rabbitmq/PACKAGES/rabbitmq-server-windows-3.8.10+1443.gf95d2ba.zip
makensis -V2 rabbitmq-3.8.10+1443.gf95d2ba.nsi
!insertmacro: macro "_If" requires 4 parameter(s), passed 3!
Error in script "rabbitmq-3.8.10+1443.gf95d2ba.nsi" on line 389 -- aborting creation process
make[2]: *** [Makefile:19: dist] Error 1
make[2]: Leaving directory '/home/lbakken/development/rabbitmq/rabbitmq-packaging/windows-exe'
make[1]: *** [Makefile:102: package-windows] Error 2
make[1]: Leaving directory '/home/lbakken/development/rabbitmq/rabbitmq-packaging'
make: *** [Makefile:392: package-windows] Error 2
```

NSIS version:

```
lbakken@shostakovich ~/development/rabbitmq/rabbitmq-server (master =)
$ pacman -Qi nsis
Name            : nsis
Version         : 3.08-1
Description     : A professional open source system to create Windows installers
Architecture    : x86_64
URL             : http://nsis.sourceforge.net
Licenses        : custom:zlib
Groups          : None
Provides        : None
Depends On      : gcc-libs
Optional Deps   : None
Required By     : None
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 8.57 MiB
Packager        : Martchus <martchus@gmx.net>
Build Date      : Thu 07 Oct 2021 05:58:43 AM PDT
Install Date    : Thu 07 Oct 2021 12:55:23 PM PDT
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : SHA-256 Sum
```

**Comment by inikulshin** [comment]
@lukebakken I couldn't decide what should be the Erlang version upper limit, and finally left it invalid :cry:

I fixed it now to support `ERLANG_TO_VERSION` macro with no value and treat it like no upper limit. **I also left it empty,** but I'm not sure it should be used like that.

I think, it should exactly match the [RabbitMQ and Erlang/OTP Compatibility Matrix](https://www.rabbitmq.com/which-erlang.html#compatibility-matrix), and it also must be updated together with this document (every new supported Erlang/OTP version).

So, to support Erlang/OTP up to `25.0.*` (`erts-13.0.*`) `ERLANG_TO_VERSION` must be set to `13.1` as it is used as **not included upper bound**.

**Comment by inikulshin** [comment]
@lukebakken you may also prefer to keep `ERLANG_TO_VERSION` empty for `3.10.x` versions of RabbitMQ until there will be unsupported Erlang/OTP version, and only then update relevant `3.10.*` tags (and rebuild the installers)...  It seems a worse solution for me.

File: windows-exe/rabbitmq_nsi.in [hunk 1]
```diff
@@ -9,9 +11,16 @@
 
 !addplugindir plugins
 
+!define ERLANG_FROM_VERSION 11.1.4  ; https://www.erlang.org/patches/otp-23.2#erts-11.1.4
+!define ERLANG_TO_VERSION   ""      ; https://www.erlang.org/patches/otp-24.3.4#erts-12.3.2
```
[the remark this answers is not in the document]
Comment (inikulshin) [reply windows-exe/rabbitmq_nsi.in]: See comments

**Comment by michaelklishin** [comment]
@inikulshin that's actually fine and what we ended up doing for various reasons for our RPM and Debian packages.

I have a minor naming request before this can go in, through. Can we rename `ERLANG_FROM_VERSION` and `ERLANG_TO_VERSION` to `ERLANG_MIN_VERSION` and `ERLANG_MAX_VERSION`, respectively? Thank you.

**Comment by inikulshin** [comment]
@michaelklishin ok.

But, note that the valid range is `[`ERLANG_MIN_VERSION, ERLANG_MAX_VERSION`)` (like C++'s [begin, end) range)
I did it to use ERLANG_MAX_VERSION=13.1 for enabling `erts-13.0.*`. Updating ERLANG_MAX_VERSION is required only for new minor erts.

**Comment by michaelklishin** [comment]
@inikulshin we don't use ERTS versions in any other package. I'd much prefer actual OTP release versions, since that's what you see or pin when you provision Erlang.

**Comment by inikulshin** [comment]
@michaelklishin initially, I refactored and fixed Erlang version compare code just because I saw a very old `8.3`, which is **erts** version.
It's not something I introduced or broke. But if I already here... enjoy 😄 
\* I extracted it from the same registry, that is used for ERLANG_HOME, the best I found.

**Comment by inikulshin** [comment]
@michaelklishin 
Btw, does [`23.2`,`25.0`] from https://www.rabbitmq.com/which-erlang.html#compatibility-matrix actually means [`23.2.0`,`25.0.*`] (or [`23.2.0`,`25.1.0`), like I implemented)?

**Comment by michaelklishin** [comment]
@inikulshin It means `[23.2.*, 25.*]`, at least these days when it's rare to encounter incompatibilities between minor Erlang versions.

Very soon it will be `[23.3.*, 25.*]` and in about a month, `[24.*, 25.*]

**Comment by lukebakken** [comment]
Thanks, I just built an installer successfully. I'll begin testing in more depth tomorrow.

**Review by lukebakken — CHANGES_REQUESTED** [review]
@inikulshin there seems to be an issue with extracting to the `plugins/` sub-directory of the installation directory. Please see the following transcript of env var I set (`RABBITMQ_BASE`) and installation:

[error-no-plugins.txt](https://github.com/rabbitmq/rabbitmq-packaging/files/9134445/error-no-plugins.txt)

**Comment by lukebakken** [comment]
My guess is that this commit will fix the plugins issue - https://github.com/rabbitmq/rabbitmq-packaging/commit/0f5758b16f08c64d2fa1e416879288e9acc79259

@inikulshin FYI the way you have forked this repository and patched `master` (rather than a dedicated feature/topic branch) makes it impossible for me to push fixes like the above to your fork.

In the future, our preferred workflow is:
* User forks a repo in https://github.com/rabbitmq to their account, _preserving the repo's name_
* User creates a topic branch in their fork for their changes.
* User opens a pull request based on that topic branch.

If I were to then need to make minor changes, I could just push them directly to your fork. As it stands, you have to apply my commit https://github.com/rabbitmq/rabbitmq-packaging/commit/0f5758b16f08c64d2fa1e416879288e9acc79259 manually or cherry-pick it.

**Review by lukebakken — CHANGES_REQUESTED** [review]
I tried `/NOSERVICERUN` but the service was started anyway:

```
C:\Users\bakkenl\rmq-server>set | findstr RABBIT
RABBITMQ_BASE=C:\ProgramData\RabbitMQ Data

C:\Users\bakkenl\rmq-server>".\rabbitmq-server-3.8.10+1497.g4de5550.exe" /NOSERVICERUN /NOSTARTMENU

C:\Users\bakkenl\rmq-server>sc query rabbitmq

SERVICE_NAME: rabbitmq
        TYPE               : 10  WIN32_OWN_PROCESS
        STATE              : 4  RUNNING
                                (STOPPABLE, NOT_PAUSABLE, ACCEPTS_SHUTDOWN)
        WIN32_EXIT_CODE    : 0  (0x0)
        SERVICE_EXIT_CODE  : 0  (0x0)
        CHECKPOINT         : 0x0
        WAIT_HINT          : 0x0
```

Given that the installer would never have worked due to [this bug](https://github.com/rabbitmq/rabbitmq-packaging/pull/15#issuecomment-1188381772) (since fixed), I'd like to ask that you thoroughly re-test your changes before requesting another review from me. Thanks.

**Comment by inikulshin** [comment]
@lukebakken I'm sorry, I tested all new flags combinations, but only together with `/S` flag.
It's implied, that they are for silent mode - why someone needs to type any flags in command line, if he can toggle those options interactively with a single click?

But you are definitely right: those flags should be either disallowed in interactive mode, or fully supported.

**Comment by michaelklishin** [comment]
I vote for "supported". CLI users like their CLI flags and do not like clicking around ;)

**Comment by michaelklishin** [comment]
We have basic package operation tests on Concourse. They are not hooked to this repository but will inevitably catch anything obviously broken.

**Comment by michaelklishin** [comment]
@inikulshin as long as the default behavior is the same, I don't think the newly added/respected environment variables can do any harm.

**Comment by lukebakken** [comment]
Thanks @inikulshin. I just returned from PTO and will re-test this soon.

**Review by lukebakken — APPROVED** [review]
Thank you! And thank you for your patience.

**Comment by michaelklishin** [comment]
This will first ship in `3.10.8` and `3.9.22`.

@inikulshin thank you and it would be great if you could verify — once we have alpha builds of the above versions up on the [Releases page of this repo](https://github.com/rabbitmq/rabbitmq-server-binaries-dev/releases) — that it does not affect upgrades. When default data directory
locations are made configurable, this is always really scary to release.

**Review by michaelklishin — APPROVED** [review]

**Comment by michaelklishin** [comment]
This repo does not have Mergify enabled, I will backport manually.

**Comment by michaelklishin** [comment]
Backports are all done.

**Comment by inikulshin** [comment]
@lukebakken @michaelklishin thank you for your patience :)

And also for your commitment to the quality. Sometimes, as a reviewer, I feel uncomfortable to request more and more minor changes from my colleagues, but this is the way to do it right, thanks.

I enjoyed it :)

The annoying thing that with all my bugs and slow responses (sometimes days for one line fix), I missed the end of life of 3.8.*,
And now it is not clear when we'll migrate to the newer version in RabbitMQ.

But I'll try to run all our tests locally soon.

**Comment by michaelklishin** [comment]
@inikulshin this is likely related. I have noticed that Windows package builds have started failing with

```
makensis -V2 rabbitmq-3.11.0-alpha.34.nsi

!insertmacro: macro "FUNCTION_STRING_StrTok" requires 0 parameter(s), passed 5!

Error in script "rabbitmq-3.11.0-alpha.34.nsi" on line 388 -- aborting creation process
```

Do you know what'd be the right way to address it?

**Comment by michaelklishin** [comment]
We will revert this PR until the above issue is addressed as a Windows package build failure  blocks our pipelines.

**Comment by inikulshin** [comment]
@michaelklishin which makensis version are you using for this build?
I think we all tested it with relatively new one, but maybe on your build machine it is still old?

**Comment by michaelklishin** [comment]
It seems to be `3.04` from the `nsis` package in [Debian Buster](https://packages.debian.org/search?suite=all&section=all&arch=any&searchon=names&keywords=nsis).

**Comment by lukebakken** [comment]
Ah, I tested with `3.08`:

```
$ makensis -VERSION
v3.08
```

**Comment by michaelklishin** [comment]
That is only available in Debian Bookworm. Upgrading our pipelines to use it may not be possible in the short term.

**Comment by michaelklishin** [comment]
@lukebakken @inikulshin would it be possible to polish this to be `3.04`-compatible? Would `3.06` be more pleasant to work with? Migrating to Debian Bullseye may be an option.

**Comment by lukebakken** [comment]
I'll see if I can modify this PR for nsis 3.04.

**Comment by lukebakken** [comment]
@inikulshin if you'd like to build a docker image to build the windows installer using pretty much the same process we do, check this out - https://github.com/lukebakken/rabbitmq-packaging-15

**Comment by lukebakken** [comment]
@michaelklishin, good news, updating to Bullseye appears to resolve the issue since NSIS `v3.06.1-1` is available.

```
make[3]: Leaving directory '/root/rabbitmq-packaging/windows/rabbitmq-server-3.8.10+1588.gff8dfda'
cp -a README-etc rabbitmq_server-3.8.10+1588.gff8dfda/etc/README.txt
zip -q -r rabbitmq-server-windows-3.8.10+1588.gff8dfda.zip rabbitmq_server-3.8.10+1588.gff8dfda
if test "/root/PACKAGES"; then \
        mkdir -p "/root/PACKAGES"; \
        mv rabbitmq-server-windows-3.8.10+1588.gff8dfda.zip "/root/PACKAGES"; \
fi
rm -rf rabbitmq-server-* rabbitmq_server-*
make[2]: Leaving directory '/root/rabbitmq-packaging/windows'
make -C windows-exe SOURCE_DIST_FILE="/root/rabbitmq-server/PACKAGES/rabbitmq-server-3.8.10+1588.gff8dfda.tar.xz" PACKAGES_DIR="/root/PACKAGES" SIGNING_KEY="" all clean
make[2]: Entering directory '/root/rabbitmq-packaging/windows-exe'
sed \
    -e 's|%%VERSION%%|3.8.10+1588.gff8dfda|' \
    -e 's|%%PRODUCT_VERSION%%|3.8.10.0|' \
    rabbitmq_nsi.in > rabbitmq-3.8.10+1588.gff8dfda.nsi
unzip -q /root/PACKAGES/rabbitmq-server-windows-3.8.10+1588.gff8dfda.zip
makensis -V2 rabbitmq-3.8.10+1588.gff8dfda.nsi
rm -rf rabbitmq-*.nsi rabbitmq_server-* rabbitmq-server-*.exe
make[2]: Leaving directory '/root/rabbitmq-packaging/windows-exe'
make[1]: Leaving directory '/root/rabbitmq-packaging'
root@b040a32e68f0:~/rabbitmq-server# makensis -VERSION;echo
v3.06.1-1
```

I'll open a new PR with @inikulshin's changes.

**Comment by lukebakken** [comment]
@michaelklishin all set with two new PRs. @inikulshin 's changes are in `master` and `v3.11.x`. I just needed to revert the reverts in `3.10.x` and `3.9.x`.

**Pull request merged**